deeppavlov/models/ranking/matching_models/tf_base_matching_model.py

```diff
@@ -0,0 +1,205 @@
+                            batch_buffer_context_len[i * self.batch_size:(i + 1) * self.batch_size]),
+                        self.response_ph: np.array(
+                            batch_buffer_response[i * self.batch_size:(i + 1) * self.batch_size]),
+                        self.response_len_ph: np.array(
+                            batch_buffer_response_len[i * self.batch_size:(i + 1) * self.batch_size])
+                    }
+                    yp = self.sess.run(self.y_pred, feed_dict=feed_dict)
+                    y_pred += list(yp[:, 1])
+                break
+        y_pred = np.asarray(y_pred)
+        if len(response_sentences) > 1:
+            y_pred = np.reshape(y_pred, (j, len(response_sentences)))  # reshape to [batch_size, 10]
+        return y_pred
+
+    # load() and save() are inherited from TFModel
+
+    def train_on_batch(self, x: List[np.ndarray], y: List[int]) -> float:
+        """
+        This method is called by trainer to make one training step on one batch.
+
+        :param x: generator that returns
+                  list of ndarray - words of all sentences represented as integers,
+                  with shape: (number_of_context_turns + 1, max_number_of_words_in_a_sentence)
+        :param y: tuple of labels, with shape: (batch_size, )
+        :return: value of loss function on batch
+        """
+        batch_buffer_context = []       # [batch_size, 10, 50]
+        batch_buffer_context_len = []   # [batch_size, 10]
+        batch_buffer_response = []      # [batch_size, 50]
+        batch_buffer_response_len = []  # [batch_size]
+        j = 0
+        while True:
+            try:
+                sample = next(x)
+                j += 1
+                context_sentences = sample[:self.num_context_turns]
+                response_sentences = sample[self.num_context_turns:]
+
+                # format model inputs
+                # word indices
+                batch_buffer_context += [context_sentences for sent in response_sentences]
+                batch_buffer_response += [response_sentence for response_sentence in response_sentences]
+                # lens of sentences
+                lens = []
+                for context in [context_sentences for sent in response_sentences]:
+                    context_sentences_lens = []
+                    for sent in context:
+                        context_sentences_lens.append(len(sent[sent != 0]))
+                    lens.append(context_sentences_lens)
+                batch_buffer_context_len += lens
+
+                lens = []
+                for context in [response_sentence for response_sentence in response_sentences]:
+                    lens.append(len(context[context != 0]))
+                batch_buffer_response_len += lens
+
+                if len(batch_buffer_context) >= self.batch_size:
+                    for i in range(len(batch_buffer_context) // self.batch_size):
+                        feed_dict = {
```

Comment:
https://github.com/deepmipt/DeepPavlov/blob/14266e0627eb8675ea51da21697a007d39a9dfba/deeppavlov/models/ranking/matching_models/tf_base_matching_model.py#L165-L175 This is duplication of the `__call__` method code.